File: infer/tests/codetoanalyze/java/pulse/taint/Strings.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package codetoanalyze.java.pulse;

import java.util.Formatter;

/**
 * a lot of tainted values are strings, so propagation through StringBuilder's and the like is very
 * important.
 */
public class Strings {

  void viaStringBuilderSugarBad() {
    Object source = InferTaint.inferSecretSource();
    InferTaint.inferSensitiveSink(source + "");
  }

  void viaStringBuilderBad() {
    Object source = InferTaint.inferSecretSource();
    StringBuilder builder = new StringBuilder();
    InferTaint.inferSensitiveSink(builder.append(source).append("").toString());
  }

  void viaStringBuilderIgnoreReturnBad() {
    Object source = InferTaint.inferSecretSource();
    StringBuilder builder = new StringBuilder();
    // builder should be tainted after this call even though we ignore the return value
    builder.append(source);
    InferTaint.inferSensitiveSink(builder.toString());
  }

  void viaStringBufferBad() {
    Object source = InferTaint.inferSecretSource();
    StringBuffer buffer = new StringBuffer();
    InferTaint.inferSensitiveSink(buffer.append("").append(source).toString());
  }

  void viaStringBufferIgnoreReturnBad() {
    Object source = InferTaint.inferSecretSource();
    StringBuffer buffer = new StringBuffer();
    buffer.append(source);
    InferTaint.inferSensitiveSink(buffer.toString());
  }

  void viaFormatterBad() {
    Object source = InferTaint.inferSecretSource();
    Formatter formatter = new Formatter();
    InferTaint.inferSensitiveSink(formatter.format("%s", source).toString());
  }

  void viaFormatterIgnoreReturnBad() {
    Object source = InferTaint.inferSecretSource();
    Formatter formatter = new Formatter();
    formatter.format("%s", source);
    InferTaint.inferSensitiveSink(formatter.toString());
  }

  void viaStringFormatVarArgsDirectBad() {
    Object source = InferTaint.inferSecretSource();
    String tainted = String.format("%s%s", "hi", source);
    InferTaint.inferSensitiveSink(tainted);
  }

  void viaStringFormatVarArgsIndirect(Object param) {
    String tainted = String.format("%s%s", "hi", param);
    InferTaint.inferSensitiveSink(tainted);
  }

  void viaStringFormatVarArgsIndirectBad() {
    viaStringFormatVarArgsIndirect(InferTaint.inferSecretSource());
  }
}

File: infer/tests/codetoanalyze/kotlin/pulse/taint/Strings.kt
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
package codetoanalyze.kotlin.pulse

import codetoanalyze.kotlin.pulse.InferTaint.inferSecretSource
import codetoanalyze.kotlin.pulse.InferTaint.inferSensitiveSink
import java.util.*

/**
 * a lot of tainted values are strings, so propagation through StringBuilder's and the like is very
 * important.
 */
class Strings {

  fun viaStringBuilderSugarBad() {
    val source = inferSecretSource()
    inferSensitiveSink(source.toString() + "")
  }

  fun viaStringBuilderBad() {
    val source = inferSecretSource()
    val builder = StringBuilder()
    inferSensitiveSink(builder.append(source).append("").toString())
  }

  fun viaStringBuilderIgnoreReturnBad() {
    val source = inferSecretSource()
    val builder = StringBuilder()
    // builder should be tainted after this call even though we ignore the return value
    builder.append(source)
    inferSensitiveSink(builder.toString())
  }

  fun viaStringBufferBad() {
    val source = inferSecretSource()
    val buffer = StringBuffer()
    inferSensitiveSink(buffer.append("").append(source).toString())
  }

  fun viaStringBufferIgnoreReturnBad() {
    val source = inferSecretSource()
    val buffer = StringBuffer()
    buffer.append(source)
    inferSensitiveSink(buffer.toString())
  }

  fun viaFormatterBad() {
    val source = inferSecretSource()
    val formatter = Formatter()
    inferSensitiveSink(formatter.format("%s", source).toString())
  }

  fun viaFormatterIgnoreReturnBad() {
    val source = inferSecretSource()
    val formatter = Formatter()
    formatter.format("%s", source)
    inferSensitiveSink(formatter.toString())
  }

  fun viaStringFormatVarArgsDirectBad() {
    val source = inferSecretSource()
    val tainted = String.format("%s%s", "hi", source)
    inferSensitiveSink(tainted)
  }

  fun viaStringFormatVarArgsIndirect(param: Any?) {
    val tainted = String.format("%s%s", "hi", param)
    inferSensitiveSink(tainted)
  }

  fun viaStringFormatVarArgsIndirectBad() {
    viaStringFormatVarArgsIndirect(inferSecretSource())
  }
}
